### PR TITLE
Add device handler

### LINKED ISSUE #602: Add device handler for REST methods/status checking for devices

### PR DESCRIPTION
**Pull request by nagem**
Closes #602

### Review Checklist

- Tests were added to cover all code changes
- Documentation was added / updated
- Code and tests follow standards in CONTRIBUTING.md

### CODE REVIEW

**Comment by nagem** [comment]
Opening PR for review now while I finish up documentation/json schema/minor tests.

File: api/handlers/devicehandler.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+log = config.log
+
+Status = util.Enum('Origin', {
+    'ok':       'ok',       # Device's last seen time is shorter than expected interval for checkin, no errors listed.
+    'missing':  'missing',  # Device's last seen time is longer than the expected interval for checkin, but no errors listed.
+    'error':    'error' ,   # Device has errors listed.
+    'unknown':  'unknown'   # Device did not set an expected checkin interval.
+})
+
+
```
Comment (nagem) [inline api/handlers/devicehandler.py]: Let me know what you think of these simple permissions checking decorators. I thought I'd experiment with this change and if we like them, I'll move them to the auth package instead.

File: api/handlers/devicehandler.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        #validators.validate_data(payload, 'device.json', 'input', 'PUT', optional=True)
+
+        result = self.storage.update_el(device_id, payload)
+        if result.modified_count == 1:
+            return {'modified': result.modified_count}
+        else:
+            raise APINotFoundException('Device with id {} not found, state not updated'.format(device_id))
+
+    @require_superuser
+    def delete(self, device_id):
```
Comment (nagem) [inline api/handlers/devicehandler.py]: Might be useful for cleanup of old devices. I'll flesh this out if we think there is a need.

File: api/handlers/devicehandler.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    @require_login
+    def get(self, device_id):
+        return self.storage.get_container(device_id)
+
+    @require_login
+    def get_all(self):
+        return self.storage.get_all_el(None, None, None)
+
+    @require_drone
+    def get_self(self):
```
Comment (nagem) [inline api/handlers/devicehandler.py]: This can double as a heartbeat for devices that don't already check in at a constant interval.

**Comment by nagem** [comment]
>Coverage increased (+0.3%)

It's not much but it's something! 🎉

File: api/handlers/devicehandler.py (hunk 1 above)
Comment (kofalt) [inline api/handlers/devicehandler.py]: If a checkin interval is implicitly required, why not explicitly require it and ditch the `unknown` state?

File: api/handlers/devicehandler.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    Will also ensure superuser, which is implied with a drone request.
+    """
+    def check_drone(self, *args, **kwargs):
+        if self.origin.get('type', '') != Origin.device:
+            raise APIPermissionException('Drone request required.')
+        if not self.superuser_request:
+            raise APIPermissionException('Superuser required.')
+        return handler_method(self, *args, **kwargs)
+    return check_drone
```
Comment (kofalt) [inline api/handlers/devicehandler.py]: Why not take these three closure generators and move them to util?

This looks like a decent pattern you're starting, might as well use it everywhere (eventually).

File: api/handlers/devicehandler.py (hunk 3 above)
Comment (kofalt) [inline api/handlers/devicehandler.py]: Not sure I understand your comment, could you elaborate?

File: api/handlers/devicehandler.py (hunk 2 above)
Comment (kofalt) [inline api/handlers/devicehandler.py]: :+1:

File: api/handlers/devicehandler.py (hunk 1 above)
Comment (kofalt) [inline api/handlers/devicehandler.py]: Also, change `Origin` to `Status`, right?

File: api/base.py [hunk 5]
```diff
@@ -241,9 +241,10 @@ def set_origin(self, drone_request):
                 }, {
                     '$set': {
                         '_id': self.origin['id'],
-                        'last-seen': datetime.datetime.utcnow(),
+                        'last_seen': datetime.datetime.utcnow(),
                         'method': self.origin['method'],
-                        'name': self.origin['name']
+                        'name': self.origin['name'],
+                        'errors': None # Reset errors list if device checks in
```
Comment (kofalt) [inline api/base.py]: Can we please set `errors` to an empty array instead of null? Consistent typing.

Comment (nagem) [inline api/base.py]: Sure.

File: api/handlers/devicehandler.py (hunk 4 above)
Comment (nagem) [inline api/handlers/devicehandler.py]: I'll move them to  `auth`.

**Comment by nagem** [comment]
Made changes noted in review, will merge when CI is green.

**Comment by nagem** [comment]
I don't know why coverage went down .3% when all I did was move some code around between commits. Oh well, I'm going to merge anyway.

**Pull request merged**